File: Two Pointer/Product of Array Except Self.py

# naive solution O(n^2) (Not optimal)
# def productExceptSelf(nums):
#     result = []
#     for i in range(len(nums)):
#         pro = 1
#         for j in range(len(nums)):
#             if i != j:
#                 pro *= nums[j]
#         result.append(pro)
#     return result


def productExceptSelf(nums):
    n = len(nums)
    result = [1]*n
    start = 0
    end = n-1
    left_product = 1
    right_product = 1
    while start < n and end >= 0:
      result[start]*=left_product
      result[end]*=right_product
      left_product*=nums[start]
      right_product*=nums[end]
      start+=1
      end-=1
    return result
      
x= [-3,3,2,-1,5]
print(productExceptSelf(x))

File: Two Pointer/Product of Array Except Self.go
package TwoPointer

//  naive solution O(n^2) (Not optimal)
// func productExceptSelf(nums[]int) []int {
// 	n := len(nums)
// 	result := make([]int, n)
// 	for i := 0; i < n; i++ {
// 		pro := 1
// 		for j := 0; j < n; j++ {
// 			if i != j {
// 				pro *= nums[j]
// 			}
// 		}
// 		result[i] = pro
// 	}
// 	return result
// }




// optimal solution O(n)
func productExceptSelf(nums []int) []int {
	n := len(nums)
	res := make([]int,n)
	for i:= range(res){
		res[i] = 1
	} 

	start , end := 0 , n-1
	left_proudct ,right_proudct := 1,1
	for start < n && end >= 0 {
		res[start]*=left_proudct
		res[end]*=right_proudct
		left_proudct *= nums[start]
		right_proudct *= nums[end]
		start++
		end--
	} 
	return res
}